// eslint-plugin-import/lib/rules/exports-last.a9fc3671b7e4.js
'use strict';var _docsUrl = require('../docsUrl');var _docsUrl2 = _interopRequireDefault(_docsUrl);function _interopRequireDefault(obj) {return obj && obj.__esModule ? obj : { 'default': obj };}

function isNonExportStatement(_ref) {var type = _ref.type;
  return type !== 'ExportDefaultDeclaration' &&
  type !== 'ExportNamedDeclaration' &&
  type !== 'ExportAllDeclaration';
}

module.exports = {
  meta: {
    type: 'suggestion',
    docs: {
      category: 'Style guide',
      description: 'Ensure all exports appear after other statements.',
      url: (0, _docsUrl2['default'])('exports-last') },

    schema: [] },


  create: function () {function create(context) {
      return {
        Program: function () {function Program(_ref2) {var body = _ref2.body;
            var lastNonExportStatementIndex = body.reduce(function () {function findLastIndex(acc, item, index) {
                if (isNonExportStatement(item)) {
                  return index;
                }
                return acc;
              }return findLastIndex;}(), -1);

            if (lastNonExportStatementIndex !== -1) {
              body.slice(0, lastNonExportStatementIndex).forEach(function () {function checkNonExport(node) {
                  if (!isNonExportStatement(node)) {
                    context.report({
                      node: node,
                      message: 'Export statements should appear at the end of the file' });

                  }
                }return checkNonExport;}());
            }
          }return Program;}() };

    }return create;}() };
